plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerServerEvaluationStrategy.java

```diff
@@ -0,0 +1,82 @@
+package org.eclipse.che.plugin.docker.machine;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.che.api.machine.server.model.impl.ServerImpl;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
+import org.eclipse.che.plugin.docker.client.json.PortBinding;
+
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+
+/**
+ * Represents a server evaluation strategy for the configuration where the workspace server and
+ * workspace containers are running on the same Docker network. Calling
+ * {@link ServerEvaluationStrategy#getServers(ContainerInfo, String, Map)} will return a completed
+ * {@link ServerImpl} with internal addresses set to the container's address within the internal
+ * docker network. If the server cannot directly ping the container, communication will fail.
+ *
+ * <p>The addresses used for external address can be overridden via the property {@code che.docker.ip.external}.
+ *
+ * @author Angel Misevski <amisevsk@redhat.com>
+ * @see ServerEvaluationStrategy
+ */
+public class LocalDockerServerEvaluationStrategy extends ServerEvaluationStrategy {
+
+    @Inject
+    public LocalDockerServerEvaluationStrategy (@Nullable @Named("che.docker.ip") String internalAddress,
+                                                @Nullable @Named("che.docker.ip.external") String externalAddress) {
+        super(internalAddress, externalAddress);
+    }
+
+    @Override
+    protected Map<String, String> getInternalAddressesAndPorts(ContainerInfo containerInfo, String internalHost) {
+        String internalAddressContainer = containerInfo.getNetworkSettings().getIpAddress();
+
+        String internalAddress;
+        boolean useExposedPorts = true;
+        if (internalAddressContainer != null && !internalAddressContainer.isEmpty()) {
+            internalAddress = internalAddressContainer;
+        } else {
+            internalAddress = internalHost;
```

Comment:
Internal Host is a holdover from before -- it ultimately comes from `DockerNode#getHost()`. Prior to these changes, this value was used only when running Che Server as a standalone tomcat server. I've left it here as a fallback, but it could probably be removed from LocalDockerServerEvaluationStrategy altogether. It has lowest precedence because that was the way it was used before: [it's the same as internalHostnameAssisted here](https://github.com/eclipse/che/blob/11b3b9de67cd9f7c40e3641b748e97af8ae3c89a/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/LocalDockerInstanceRuntimeInfo.java#L93). Note that when running Che using che-launcher, the env variable `CHE_DOCKER_MACHINE_HOST_INTERNAL` is set to [the value defined here](https://github.com/eclipse/che/blob/11b3b9de67cd9f7c40e3641b748e97af8ae3c89a/dockerfiles/launcher/launcher_funcs.sh#L273)

The code that defines internalHost is in [DockerConnectorConfiguration](https://github.com/eclipse/che/blob/11b3b9de67cd9f7c40e3641b748e97af8ae3c89a/plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/DockerConnectorConfiguration.java#L153).